### PR TITLE
Configurable Keycloak

### PR DESCRIPTION
**Pull request by ibacher**
I apologise for the size of this PR. It does several things (adds OpenMRS code formatting, renames the project), but the main thing is the implementation of the configuration options. As you can see, I haven't written tests (it's hard to write tests for an unfamiliar environment) and I still haven't gotten this to run in my environment. I'm hoping you can give this a try on your machine (you don't necessarily need to merge this. From the command line, you should be able to checkout this branch by doing:

```
git checkout -b config origin/config
```

And then:

```
mvn clean install
```

To run things. Let me know how this goes.

PS I should say, it's hard to write *correct* tests for an unfamiliar environment... I could write tests for this, but they would simply assert that it does what I think it should do rather than that it does what Keycloak needs it to do.

### CODE REVIEW

**Comment by VaishSiddharth** [comment]
@ibacher Ok I am trying to run this will soon get back to you. Thanks for your work :+1:

File: src/main/java/org/openmrs/keycloak/provider/OpenmrsAuthenticatorProviderFactory.java [hunk 1]
```diff
@@ -26,164 +40,142 @@
...
+		        .type(ProviderConfigProperty.PASSWORD).secret(true).add().build();
 	}
-
+	
 	@Override
 	public OpenmrsAuthenticator create(KeycloakSession keycloakSession, ComponentModel model) {
 		MultivaluedHashMap<String, String> config = model.getConfig();
-		EntityManagerFactory emf = new HibernatePersistenceProvider()
-				.createContainerEntityManagerFactory(new PersistenceUnitInfoImpl(), ImmutableMap.<String, Object>builder()
-						.put(AvailableSettings.JPA_JDBC_DRIVER, "com.mysql.jdbc.Driver")
```
Comment (VaishSiddharth) [inline src/main/java/org/openmrs/keycloak/provider/OpenmrsAuthenticatorProviderFactory.java]: @ibacher Shouldn't this be
```
put(AvailableSettings.JPA_JDBC_DRIVER, "com.mysql.cj.jdbc.Driver")
```

Comment (ibacher) [reply src/main/java/org/openmrs/keycloak/provider/OpenmrsAuthenticatorProviderFactory.java]: Depends on which version of the MySQL connector is running. You’re right if we’re running 8.0 or newer, but my thought was we might be able to leverage the version bundled with Keycloak, which uses connector 5.1 where this is the right class name.

Comment (VaishSiddharth) [reply src/main/java/org/openmrs/keycloak/provider/OpenmrsAuthenticatorProviderFactory.java]: @ibacher I tried running the server with both the versions and I still get this error

```
Uncaught server error: java.lang.NoClassDefFoundError: org/hibernate/jpa/HibernatePersistenceProvider
```

Any idea?

**Comment by ibacher** [comment]
Alright! Since this wasn't nearly complex enough already, I've added some changes to package things up as an [EAR](https://en.wikipedia.org/wiki/EAR_(file_format)) rather than just a JAR. The advantage of an EAR, unlike a JAR, is that it can contain all of it's dependencies or declare other dependencies it needs. It's similar to OpenMRS's OMOD format, but much less custom and hackish (OMODs pre-date EARs and we've never converted). Downside: It's been a long time since I've written an EAR, so I'm not entirely certain this (simple) setup is all that's needed. Hopefully things have been simplified enough that this is sufficient.

Also hopefully, this solves the "ClassNotFound exception". Instead of copying the JAR, however, you'll want to copy the EAR. It should be deployable in the same place.

**Comment by VaishSiddharth** [comment]
@ibacher Ok I'll read the code and get back to you after trying it. Thanks again 😊

**Pull request closed without merging**